tools/mlock.ml

```diff
@@ -383,17 +383,23 @@ module Top(O:Config)(Out:OutTests.S) = struct
   | Expand -> List.map expand_params
   | Once -> fun pss -> pss
 
-  let tr_parsed tr_ins name t = match t.extra_data with
-  | CExtra pss ->
-      changed := false ;
-      let prog =
-        List.map
-          (fun (i,code) -> i,List.map (CBase.pseudo_map tr_ins) code)
-          t.prog in
-      if not !changed then not_changed name ;
-      let extra_data = CExtra (tr_params pss) in
-      { t with extra_data; prog;}
-  | NoExtra|BellExtra _ -> assert false
+  let tr_parsed tr_ins name t =
+    let extra = t.extra_data in
+    let t_list = List.map
+      (function
+       | CExtra pss ->
+          changed := false ;
+          let prog =
+          List.map
+             (fun (i,code) -> i,List.map (CBase.pseudo_map tr_ins) code)
+             t.prog in
+          if not !changed then not_changed name ;
+          let extra_data = [CExtra (tr_params pss)] in
+           { t with extra_data; prog;}
+       | NoExtra
+       | BellExtra _ -> assert false)
+      extra in
+      List.hd (t_list)
```

Comment:
If I understand this code correctly, the original code computes a new program '{t with prog = tr_prog t.prog; extra_data = tr_data t.extra_data;}', although in a slightly contrived manner.  In other words the original code could  (should) have been written as follows:
```OCaml
  let tr_prog prog =
     changed := false ;
     let prog = 
       List.map
        (fun (i,code) -> i,List.map (CBase.pseudo_map tr_ins) code)
        prog in
     if not !changed then not_changed name ;
    prog in
  let tr_data data =
    match  data with
    | CExtra pss -> CExtra (tr_params pss)
    | NoExtra|BellExtra _ -> assert false in
  { t with prog = tr_prog t.prog; extra_data = tr_data t.extra_data; }
````
 Then the new code would be like : ` { t with prog = tr_prog t.prog; extra_data = List.map tr_data t.extra_data; }`